### ISSUE TITLE
Allure-testng attachment on test failed

### ISSUE THREAD

**Issue by kolob204** [cbeust/testng#2791]
### TestNG Version
> Note: only the latest version is supported
7.6.1

### Expected behavior
Attachment ScreenShoot appear in allure report if test failed.

### Actual behavior
Attachment ScreenShoot doesn't appear in allure report.

If i use such combination of libraries
```
JDK 11  
selenium 4.3.0  
testng 7.4.0  
allure-testng 2.18.1  
maven-surefire-plugin 3.0.0-M7  
aspectj.version 1.9.6  
```
ScreenShoot is appear in allure report (allure serve)  

But, if set testng to 7.5.0 or 7.6.1 version:  
  screenshoot not appear anymore in report. (Note: file attachment persist in allure-result folder)

### Is the issue reproducible on runner?
 
- [X ] Maven 

### Test case sample 
It's  Classic sample, and it's work on 7.4.0 version

```
public class AllureListner  implements  ITestListener  {
	@Override
	public void onTestFailure(ITestResult iTestResult) {
		Object testClass = iTestResult.getInstance();
		WebDriver driver = ((BaseTest) testClass).getWebDriver();
		saveScreenshotPNG(driver);
	}
 
	@Attachment(value = "Page screenshot", type = "image/png")
	public byte[] saveScreenshotPNG(WebDriver driver) {
        return ((TakesScreenshot) driver).getScreenshotAs(OutputType.BYTES);
    }
}

@Listeners({ AllureListner.class })
public class Test001 extends BaseTest {
//bla bla bla
}
```

**Comment by krmahadevan** [comment]
@kolob204 i think you should perhaps have the allure dev take a look at this to find out what is the mismatch. Not sure what can be done here from TestNG side.

**Comment by mstrahinski** [comment]
This thing has no relation to TestNG, I'm using TestNG + Allure myself and everything works perfect, however you need to be aware that Allure themselves are recommending you that u take the screenshot within their lifecycle(TestLifecycleListener) and NOT the ITestListener lifecycle.

You might wanna have a look on:

https://www.testautomatisierung.org/allure-lifecycle-listener-implementieren/

and

https://github.com/allure-framework/allure-java/issues/444

**Comment by anhtester** [comment]
I have the same problem and have not found a solution for TestNG 7.7.0 @kolob204

**Comment by anhtester** [comment]
I tried but still not working. @mstrahinski

**Comment by slymn00** [comment]
@anhtester You got any solution for that issue ? I got a same problem.

**Comment by kolob204** [comment]
@anhtester @slymn00

1. Create file  resources/META-INF/services/io.qameta.allure.listener.TestLifecycleListener
2. write inside this file - path to Listener (chain of packages+class name) (For examle: Listners.MyListener)
3.  create your listner like this
```
public class MyListener implements TestLifecycleListener {

	@Override
	public void beforeTestStop(TestResult result) {

		  if (result.getStatus() == Status.FAILED || result.getStatus() == Status.BROKEN) {
	            if (driver != null)
	                Allure.addAttachment(result.getName(), new ByteArrayInputStream(((TakesScreenshot) driver).getScreenshotAs(OutputType.BYTES)));
	        }
		  
	}
}
```

For more information: google it: _Java ServiceLoader mechanism (SPI)_

**Comment by slymn00** [comment]
@kolob204 Thanx a lot brother it's working.

**Comment by anhtester** [comment]
> 

Can you share the sample project you made? @slymn00

**Comment by anhtester** [comment]
> 

Can you share the sample project you made? @kolob204 

I don't know how to do META-INF/services

**Comment by slymn00** [comment]
@anhtester 
Firstly create a java class whateveryounamewant and then paste this code 
`public void beforeTestStop(TestResult result) {

        if (result.getStatus() == Status.FAILED || result.getStatus() == Status.BROKEN) {
            if (Driver.getDriver() != null)
                Allure.addAttachment(result.getName(), new ByteArrayInputStream(((TakesScreenshot) Driver.getDriver()).getScreenshotAs(OutputType.BYTES)));
        }

    }`

And my dependencies are -> 
`<dependency>
            <groupId>io.qameta.allure</groupId>
            <artifactId>allure-testng</artifactId>
            <version>2.19.0</version>
        </dependency>

        
        <dependency>
            <groupId>io.qameta.allure</groupId>
            <artifactId>allure-rest-assured</artifactId>
            <version>2.19.0</version>
        </dependency>

<dependency>
            <groupId>org.testng</groupId>
            <artifactId>testng</artifactId>
            <version>7.7.0</version>
        </dependency>`

Next step is go test package -> resources -> create directory name is META-INF -> create directory name is services-> create file name is io.qameta.allure.listener.TestLifecycleListener and go inside this file -> write your whateveryounamewant class path like this bases.utilities.TestListeners

And Final step implement your whateveryounamewant class with TestLifecycleListener. It's done sorry for my bad english.

**Comment by anhtester** [comment]
> 

Yeah, thank you so much. It works. @slymn00 

![image](https://user-images.githubusercontent.com/87883620/215047588-35c1b589-d890-4498-8ba4-2e80772d17ab.png)

**Comment by anhtester** [comment]
Hi, everyone can take a look at my project to see how to fix the error with TestNG > 7.5.0

**Comment by baev** [comment]
All the issue details & working examples can be found here https://github.com/biczomate/allure-testng7.5-attachment-example
